### PR TITLE
Update transport chapter, mention route()

### PR DESCRIPTION
**Pull request by Robinlovelace**

### CODE REVIEW

**Comment by Robinlovelace** [comment]
This is a major refactor of this chapter. Some of these changes (the switch to sfnetworks) were necessary but most were due to frustration with the previous contents. This is much better I think but it's quite a big change so feedback v. welcome. Thanks!

File: 13-transport.Rmd [hunk 1]
```diff
@@ -326,86 +329,106 @@ There are three main ways to compute routes from OD data:
...
+Trips between 2.5 and 5 km Euclidean distance (or around 3 km, below which trips are can be made by walking or scooter, to 6-8 km of route distance) have a relatively high probability of being cycled, and the maximum distance increases when trips are made by [electric bike](https://www.sciencedirect.com/science/article/pii/S0967070X21003401).
+<!-- Todo: add reference (RL 2022-07) -->
+Considering this, and the utility of a small route dataset for testing, we will compute routes for the subset of desire lines\index{desire lines} along which a high (100+) number of car trips take place that are 2.5 to 5 km in length:
 
 ```{r 13-transport-17, message=FALSE}
-desire_lines$distance = as.numeric(st_length(desire_lines))
-desire_carshort = dplyr::filter(desire_lines, car_driver > 300 & distance < 5000)
+desire_lines$distance_km = as.numeric(st_length(desire_lines)) / 1000
+desire_lines_short = desire_lines %>% 
+  filter(car_driver >= 100 & distance_km <= 5 & distance_km >= 2.5)
```
Comment (Robinlovelace) [inline 13-transport.Rmd:380]: These slightly weird subsetting criteria were selected to make the results look good FYI: different values I tried led to too many random lines far from the city centre.

File: 13-transport.Rmd [hunk 2]
```diff
@@ -326,86 +329,106 @@ There are three main ways to compute routes from OD data:
...
 ```
 
+In the code above `st_length()` calculated the length of each desire line, as described in Section \@ref(distance-relations).
+The `filter()` function from **dplyr** filtered the `desire_lines` dataset based on the criteria outlined above\index{filter operation|see{attribute!subsetting}}, as described in Section \@ref(vector-attribute-subsetting).
+The next stage is to convert these desire lines into routes.
+This is done using the publicly available OSRM service \index{routing} with the **stplanr** functions `route()` and `route_osrm()`\index{stplanr (package)} in the code chunk below:
+
 ```{r 13-transport-18, eval=FALSE}
-route_carshort = route(l = desire_carshort, route_fun = route_osrm)
+routes_short = route(l = desire_lines_short, route_fun = route_osrm,
```
Comment (Robinlovelace) [inline 13-transport.Rmd:389]: We can now set this to eval=TRUE no problem I think.

**Comment by Nowosad** [comment]
Hi @Robinlovelace -- I just read the whole chapter, made mostly minor changes, and also have some comments/suggestions:

- code/13-transport-data-gen is not reproducible
- Figure `bristol` contains red/green colors (not color blind friendly) -- can you improve it?
- `which can be obtained using **osmdata** with a command such as `bristol_region = osmdata::getbb("Bristol", format_out = "sf_polygon")`. This results in an `sf` object representing the bounds of the largest matching city region, either a rectangular polygon of the bounding box or a detailed polygonal boundary.` -- The actual result currently is a list of sf's, not an sf object...
- `summarise_if` is superseded (see `?summarise_if`), thus it would be good to replace it with `across()`
- code/13-desire.R returns an error `desire_lines_top5 = od2line(od_top5, zones_od)
Error in UseMethod("od2line", object = zones) : 
  no applicable method for 'od2line' applied to an object of class "c('tbl_df', 'tbl', 'data.frame')"`
- What is the difference between "route level" and "segment level". This is not explained...
- The routes figure could be improved (e.g., add manual legend, customize lwd, etc.)
- "For the purposes of this case study, suffice to say that the roads along which these short car journeys travel should be prioritized for investigation to understand how they can be made more conducive to sustainable transport modes." -- this sentence is hard to read...
- Update pipes (and more, if needed) in https://geocompr.github.io/geocompkg/articles/linevia.html
- The stations figure could be improved (e.g., add manual legend, customize lwd, etc.)
- Caption of figure rnetvis is incorrect
- Some todo comments still exist
- I moved exercises to a separate file. It would be good to clean them, improve if needed, and add solutions...

File: 03-attribute-operations.Rmd [hunk 3]
```diff
@@ -16,7 +16,7 @@ library(dplyr)   # tidyverse package for data frame manipulation
 library(spData)  # spatial data package introduced in Chapter 2
 ```
 
-- Also ensure you have installed the **tidyr** package, or the **tidyverse** of which it is a part, if you want to run data 'tidying' operations in Section \@ref(vec-attr-creation): 
+- Also ensure you have installed the **tidyr** package, or the **tidyverse** of which it is a part, if you want to run data 'tidying' operations in Section \@ref(vec-attr-creation).
```
Comment (Robinlovelace) [inline 03-attribute-operations.Rmd:19]: :+1:

File: 04-spatial-operations.Rmd [hunk 4]
```diff
@@ -11,7 +11,7 @@ library(dplyr)
 library(spData)
 ```
 
-- You also need to read in a couple of datasets as follows for Section \@ref(spatial-ras)
+- You also need to read in a couple of datasets as follows for Section \@ref(spatial-ras):
```
Comment (Robinlovelace) [inline 04-spatial-operations.Rmd:14]: :+1:

File: 13-transport.Rmd [hunk 5]
```diff
@@ -13,7 +13,7 @@ library(sf)
 library(tidyverse)
 library(spDataLarge)
 library(stplanr)      # geographic transport data package
-library(tmap)         # visualization package (see Chapter 8)
```
Comment (Robinlovelace) [inline 13-transport.Rmd:16]: :+1:

File: 13-transport.Rmd [hunk 6]
```diff
@@ -27,20 +27,21 @@ The effort of moving (overcoming distance) is central to the 'first law' of geog
 
 This 'law' is the basis for spatial autocorrelation\index{autocorrelation!spatial} and other key geographic concepts.
 It applies to phenomena as diverse as friendship networks and ecological diversity and can be explained by the costs of transport --- in terms of time, energy and money --- which constitute the 'friction of distance'.
-From this perspective, transport technologies are disruptive, changing geographic relationships between geographic entities including mobile humans and goods: "the purpose of transportation is to overcome space" [@rodrigue_geography_2013].
```
Comment (Robinlovelace) [inline 13-transport.Rmd:28]: :+1:

File: 13-transport.Rmd [hunk 7]
```diff
@@ -56,7 +57,7 @@ Typically, models are designed to solve a particular problem.
 For this reason, this chapter is based around a policy scenario, introduced in the next section, that asks:
 how to increase cycling in the city of Bristol?
 Chapter \@ref(location) demonstrates another application of geocomputation:
-prioritising the location of new bike shops.
+prioritizing the location of new bike shops.
```
Comment (Robinlovelace) [inline 13-transport.Rmd:60]: :+1: the last instance of British English? Hope so!

File: 13-transport.Rmd [hunk 8]
```diff
@@ -176,19 +177,19 @@ nrow(bristol_zones)
 The results of the previous code chunk shows that there are more than 10 OD pairs for every zone, meaning we will need to aggregate the origin-destination data before it is joined with `bristol_zones`, as illustrated below (origin-destination data is described in Section \@ref(desire-lines)):
 
 ```{r 13-transport-7}
-zones_attr = bristol_od %>%
```
Comment (Robinlovelace) [inline 13-transport.Rmd:178]: :+1: I need to update my RStudio defaults!

**Review by Robinlovelace — COMMENTED** [review]
Many thanks for the detailed set of changes Jakub.

**Comment by Robinlovelace** [comment]
> Hi @Robinlovelace -- I just read the whole chapter, made mostly minor changes, and also have some comments/suggestions:

Will get on the case this week. Very grateful for these detailed suggestions.

**Comment by Robinlovelace** [comment]
Hi @Nowosad, here are some responses:

- [ ] code/13-transport-data-gen is not reproducible

Good point and worth fixing, will work on that.

- [ ] Figure `bristol` contains red/green colors (not color blind friendly) -- can you improve it?

Yes, it's worth updating that figure.

- [ ] `which can be obtained using **osmdata** with a command such as `bristol_region = osmdata::getbb("Bristol", format_out = "sf_polygon")`. This results in an `sf` object representing the bounds of the largest matching city region, either a rectangular polygon of the bounding box or a detailed polygonal boundary.` -- The actual result currently is a list of sf's, not an sf object...

Good point, plan to update with a link to the documentation.

- [ ] `summarise_if` is superseded (see `?summarise_if`), thus it would be good to replace it with `across()`

👍


- [ ] code/13-desire.R returns an error `desire_lines_top5 = od2line(od_top5, zones_od)
Error in UseMethod("od2line", object = zones) : 
  no applicable method for 'od2line' applied to an object of class "c('tbl_df', 'tbl', 'data.frame')"`
  
Will explore and fix if I can reproduce this, the code seems approximately right but may need upstream changes in `stplanr` related to recent deprecating of **sp** functionality.
Eventually I want to use **od** as back-end powering this.
  
- [ ] What is the difference between "route level" and "segment level". This is not explained...

Route level data operates at the level of the entire route. For example, in the following code, `length` is a route level attribute but `quietness` is a segment level attribute. Does that seem like a good way to explain it?

```r
library(cyclestreets)
# ?journey # see example
from = c(-1.55, 53.80) # geo_code("leeds")
to = c(-1.76, 53.80) # geo_code("bradford uk")
r1 = journey(from, to)
names(r1)
#>  [1] "name"              "distances"         "time"             
#>  [4] "busynance"         "elevations"        "start_longitude"  
#>  [7] "start_latitude"    "finish_longitude"  "finish_latitude"  
#> [10] "crow_fly_distance" "event"             "whence"           
#> [13] "speed"             "itinerary"         "clientRouteId"    
#> [16] "plan"              "note"              "length"           
#> [19] "quietness"         "west"              "south"            
#> [22] "east"              "north"             "leaving"          
#> [25] "arriving"          "grammesCO2saved"   "calories"         
#> [28] "edition"           "gradient_segment"  "elevation_change" 
#> [31] "provisionName"     "geometry"          "gradient_smooth"
head(r1$length)
#> [1] 15405 15405 15405 15405 15405 15405
```

- [ ] The routes figure could be improved (e.g., add manual legend, customize lwd, etc.)

Good point, will aim to do this but not a priority.

- [ ] "For the purposes of this case study, suffice to say that the roads along which these short car journeys travel should be prioritized for investigation to understand how they can be made more conducive to sustainable transport modes." -- this sentence is hard to read...

Agreed, will fix.

- [ ] Update pipes (and more, if needed) in https://geocompr.github.io/geocompkg/articles/linevia.html

Or remove it and link to https://docs.ropensci.org/stplanr/reference/line_via.html instead.
That's my plan to keep things simple.

- [ ] The stations figure could be improved (e.g., add manual legend, customize lwd, etc.)

Good point, will aim to do this but not a priority.
Any suggestions on what should go in the manual legend?

- [ ] Caption of figure rnetvis is incorrect

Good point, will fix.

- [ ] Some todo comments still exist

👍

- [ ] I moved exercises to a separate file. It would be good to clean them, improve if needed, and add solutions...

👍

**Comment by Robinlovelace** [comment]
Just looking at fixing priority issues in order to get this merged, we can add other issues to a new issue is my thinking. In any case, not sure what the the problem with the rnetvis caption is. This seems OK to me @Nowosad : 

```
rnetvis, out.width="49%", fig.show='hold', fig.cap="Illustration of the % of car trips switching to cycling as a function of distance (left) and route network level results of this function (right).", echo=FALSE
```

**Comment by Nowosad** [comment]
> Just looking at fixing priority issues in order to get this merged, we can add other issues to a new issue is my thinking. In any case, not sure what the the problem with the rnetvis caption is. This seems OK to me @Nowosad :
> 
> ```
> rnetvis, out.width="49%", fig.show='hold', fig.cap="Illustration of the % of car trips switching to cycling as a function of distance (left) and route network level results of this function (right).", echo=FALSE
> ```

It seems to be okey now. I think the version I review had different code in this code chunk.

**Pull request merged**